src/reader.jl

```diff
@@ -1,48 +1,73 @@
 """
-    MOI.SolverInstance(mf::MOFFile, solver)
+    MOFInstance(file::String)
 
-Create a new MathOptInterface solver instance using `solver` from the MOFFile `mf`
+Read a MOF file located at `file`
+
+### Example
+
+    MOFInstance("path/to/model.mof.json")
 """
-function MOI.SolverInstance(mf::MOFFile, solver)
-    m = MOI.SolverInstance(solver)
-    v = MOI.addvariables!(m, length(mf["variables"]))
-    empty!(mf.namemap)
-    for (i, dict) in enumerate(mf["variables"])
-        mf.namemap[dict["name"]] = v[i]
-        MOI.set!(m, MOI.VariableName(), v[i], dict["name"])
-        if haskey(dict, "VariablePrimalStart")
-            MOI.set!(m, MOI.VariablePrimalStart(), v[i], dict["VariablePrimalStart"])
-        end
+function MOFInstance(file::String)
+    m = MOFInstance()
+    MOI.read!(m, file)
+    m
+end
+
+function MOI.read!(m::MOFInstance, file::String)
+    d = open(file, "r") do io
+        JSON.parse(io, dicttype=OrderedDict{String, Any})
     end
-    sense = MOI.get(mf, MOI.ObjectiveSense())
-    MOI.set!(m, MOI.ObjectiveFunction(), parse!(mf, mf["objective"]))
-    MOI.set!(m, MOI.ObjectiveSense(), sense)
-    for con in mf["constraints"]
-        c = MOI.addconstraint!(m, parse!(mf, con["function"]), parse!(mf, con["set"]))
-        MOI.set!(m, MOI.ConstraintName(), c, con["name"])
-        if haskey(con, "ConstraintPrimalStart")
-            MOI.set!(m, MOI.ConstraintPrimalStart(), c, con["ConstraintPrimalStart"])
-        end
-        if haskey(con, "ConstraintDualStart")
-            MOI.set!(m, MOI.ConstraintDualStart(), c, con["ConstraintDualStart"])
-        end
+    if length(m["variables"]) > 0
+        error("Unable to load the model from $(file). Instance is not empty!")
     end
-    m
+    src = MOFInstance(d, Dict{String, MOI.VariableReference}(), Dict{MOI.VariableReference, Int}(), Dict{UInt64, Int}(), CurrentReference(UInt64(0), UInt64(0)))
+    # delete everything in the current instance
+    empty!(m.d)
+    m.d["version"]     = "0.0"
+    m.d["sense"]       = "min"
+    m.d["variables"]   = Object[]
+    m.d["objective"]   = Object("head"=>"ScalarAffineFunction", "variables"=>String[], "coefficients"=>Float64[], "constant"=>0.0)
+    m.d["constraints"] = Object[]
+    MOI.copy!(m, src)
 end
 
-"""
-    MOI.SolverInstance(file::String, solver)
+function tryset!(dest, dict, ref, attr, str)
+    if haskey(dict, str) && MOI.canset(dest, attr, ref)
+        MOI.set!(dest, attr, ref, dict[str])
+    end
+end
 
-Create a new MathOptInterface solver instance using `solver` from the MOFFile
-located at the path `file`.
-"""
-MOI.SolverInstance(file::String, solver) = MOI.SolverInstance(MOFFile(file), solver)
+function MOI.copy!(dest::MOI.AbstractInstance, src::MOFInstance)
```

Comment:
I can see some ambiguity issues if an instance defines ``MOI.copy!(dest::MyInstance, src::MOI.AbstractInstance)``. Can we just have one generic version of `MOI.copy` defined in MOI or MOIU?

Reply:
I would prefer a convention that only `src` instances defined `copy!` and the 2nd argument was concretely typed.

But I suppose you could make one generic version that checked `canget` and `canset` for all attributes and constraints.